### ISSUE TITLE
Run HDBSCAN once, get multiple granularities of clusters

### ISSUE THREAD

**Issue by whaowhao** [scikit-learn-contrib/hdbscan#523]
Is it possible to run HDBSCAN.fit once, and get different granularities of clustering by specifying max_number_of_clusters at .predict() phase? That way, I can explore the data interactively in a hierarchical way.

**Comment by lmcinnes** [comment]
You can get multiple granularities of clustering from one run; you can't get it by specifying a number of clusters however. The current best approach in a released version involves doing something like:

```python
clusterer = hdbscan.HDBSCAN().fit(data)
a_labelling = clusterer.single_linkage_tree_.get_clusters(
            cut_distance=cut_distance,
            min_cluster_size=min_cluster_size,
        )
```
where the ``cut_distance`` is an eps value you might hand to DBSCAN. This is likely not exactly what you want, but probably close. The current master branch includes a more direct call to the ``.dbscan_clustering`` method of the main HDBSCAN object, which will do the same thing; that isn't in a released version yet though.

The alternative is to use the memory parameter (something like):
```python
labelling1 = hdbscan.HDBSCAN(memory="/tmp/hdbscan_cache", min_samples=10).fit_predict(data)
labelling2 = hdbscan.HDBSCAN(memory="/tmp/hdbscan_cache", min_samples=10, min_cluster_size=20).fit_predict(data)
labelling3 = hdbscan.HDBSCAN(memory="/tmp/hdbscan_cache", min_samples=10, min_cluster_size=50).fit_predict(data)
```
In this case a large amount of the computational cost can be amortized from the first run and cached; further runs that use varying values of ``min_cluster_size`` (but keeping ``min_samples`` fixed!) will be very cheap, with just the tree condensing and cluster extraction repeated (which is, compared to the rest of the computation, essentially instant).

**Comment by whaowhao** [comment]
Thanks for the useful answer. Regarding finding optimal eps value with DBSCAN, is there a heuristic that works well from your experience? @lmcinnes

**Comment by lmcinnes** [comment]
Not really; you can run a nearest neighbour search and look at the distribution of k-neighbour distances and see what looks useful based on that.
